From a8311aba86bb872ae5b5349448c03fc8d86d9b41 Mon Sep 17 00:00:00 2001
From: Huiwen He <hehuiwen@kylinos.cn>
Date: Wed, 18 Feb 2026 00:31:38 +0800
Subject: [PATCH 7/8] smb/client: optimize SMB1 error mapping with binary
 search

Currently, map_smb_to_linux_error() uses linear searches for both
mapping_table_ERRDOS and mapping_table_ERRSRV.

Refactor this by introducing a helper function smb1_map_to_posix() that
implements binary search(as the tables are sorted). This improves lookup
performance and reduces code duplication.

Also remove the sentinel entries from the mapping tables as they are no
longer needed with ARRAY_SIZE().

Signed-off-by: Huiwen He <hehuiwen@kylinos.cn>
Reviewed-by: ChenXiaoSong <chenxiaosong@kylinos.cn>
---
 fs/smb/client/smb1maperror.c | 58 ++++++++++++++++++------------------
 1 file changed, 29 insertions(+), 29 deletions(-)

diff --git a/fs/smb/client/smb1maperror.c b/fs/smb/client/smb1maperror.c
index be99d6eaf5cd..f2f8f5fffeb4 100644
--- a/fs/smb/client/smb1maperror.c
+++ b/fs/smb/client/smb1maperror.c
@@ -8,6 +8,7 @@
  *   Copyright (C) Andrew Tridgell 2001
  */
 
+#include <linux/bsearch.h>
 #include "cifsproto.h"
 #include "smb1proto.h"
 #include "smberr.h"
@@ -19,14 +20,24 @@ struct smb_to_posix_error {
 	int posix_code;
 };
 
+static __always_inline int smb1_posix_error_cmp(const void *key, const void *element)
+{
+	__u16 k = *(__u16 *)key;
+	const struct smb_to_posix_error *e = element;
+
+	if (k < e->smb_err)
+		return -1;
+	if (k > e->smb_err)
+		return 1;
+	return 0;
+}
+
 static const struct smb_to_posix_error mapping_table_ERRDOS[] = {
 #include "smberr_dos_map.h"
-	{0, 0}
 };
 
 static const struct smb_to_posix_error mapping_table_ERRSRV[] = {
 #include "smberr_srv_map.h"
-	{0, 0}
 };
 
 /*****************************************************************************
@@ -62,11 +73,22 @@ ntstatus_to_dos(__u32 ntstatus)
 	return NULL;
 }
 
+static int
+smb1_map_to_posix(const struct smb_to_posix_error *table,
+		  size_t num_entries, __u16 smb_err)
+{
+	const struct smb_to_posix_error *res;
+
+	res = __inline_bsearch(&smb_err, table, num_entries,
+			       sizeof(struct smb_to_posix_error),
+			      smb1_posix_error_cmp);
+	return res ? res->posix_code : -EIO;
+}
+
 int
 map_smb_to_linux_error(char *buf, bool logErr)
 {
 	struct smb_hdr *smb = (struct smb_hdr *)buf;
-	unsigned int i;
 	int rc = -EIO;	/* if transport error smb error may not be set */
 	__u8 smberrclass;
 	__u16 smberrcode;
@@ -105,34 +127,12 @@ map_smb_to_linux_error(char *buf, bool logErr)
 	/* DOS class smb error codes - map DOS */
 	if (smberrclass == ERRDOS) {
 		/* 1 byte field no need to byte reverse */
-		for (i = 0;
-		     i <
-		     sizeof(mapping_table_ERRDOS) /
-		     sizeof(struct smb_to_posix_error); i++) {
-			if (mapping_table_ERRDOS[i].smb_err == 0)
-				break;
-			else if (mapping_table_ERRDOS[i].smb_err ==
-								smberrcode) {
-				rc = mapping_table_ERRDOS[i].posix_code;
-				break;
-			}
-			/* else try next error mapping one to see if match */
-		}
+		rc = smb1_map_to_posix(mapping_table_ERRDOS,
+				       ARRAY_SIZE(mapping_table_ERRDOS), smberrcode);
 	} else if (smberrclass == ERRSRV) {
 		/* server class of error codes */
-		for (i = 0;
-		     i <
-		     sizeof(mapping_table_ERRSRV) /
-		     sizeof(struct smb_to_posix_error); i++) {
-			if (mapping_table_ERRSRV[i].smb_err == 0)
-				break;
-			else if (mapping_table_ERRSRV[i].smb_err ==
-								smberrcode) {
-				rc = mapping_table_ERRSRV[i].posix_code;
-				break;
-			}
-			/* else try next error mapping to see if match */
-		}
+		rc = smb1_map_to_posix(mapping_table_ERRSRV,
+				       ARRAY_SIZE(mapping_table_ERRSRV), smberrcode);
 	}
 	/* else ERRHRD class errors or junk  - return EIO */
 
-- 
2.53.0

